Add error message in JSON response

jkakavas opened the pull request:
When the SAML authentication is not successful, we return a SAML
Response with a status that indicates a failure. This commit adds
an error message in the REST API response along with the SAML
Response XML string so that the caller of the API can identify
that this is an unsuccessful response without needing to parse the
XML.

jkakavas committed:
Add error message in JSON response

When the SAML authentication is not successful, we return a SAML
Respnonse with a status that indicates a failure. This commit adds
an error message in the REST API response along with the SAML
Response XML string so that the caller of the API can identify
that this is an unsuccessful response without needing to parse the
XML.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Security)

jkakavas commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into return-error-init

jkakavas committed:
Merge remote-tracking branch 'origin/master' into return-error-init

tvernum commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/rest/action/RestSamlInitiateSingleSignOnAction.java:
```diff
@@ -65,6 +65,7 @@ public RestResponse buildResponse(SamlInitiateSingleSignOnResponse response, XCo
                         builder.startObject();
                         builder.field("post_url", response.getPostUrl());
                         builder.field("saml_response", response.getSamlResponse());
+                        builder.field("error", response.getError());
```
Is the contract that if `error` is `null` then it's a success and if it's non-null then there's a failure?

jkakavas replied:
yes. I contemplated adding a boolean `isSuccess` field but felt like this extra check wouldn't make much sense on the consumer side but I don't feel strongly about it if you think it makes this clearer/easier to consume.

tvernum replied:
It feels a little weird, but I can't think of anything better, unless we add the saml status as well.

jkakavas replied:
I ended up doing just that after all

jkakavas committed:
Clean up SamlAuthenticationState and add the SAML StatusCode in the
response

jkakavas committed:
Merge remote-tracking branch 'origin/master' into return-error-init

jkakavas committed:
Merge branch 'return-error-init' of github.com:jkakavas/elasticsearch into return-error-init

jkakavas committed:
fix test

jkakavas committed:
actually fix that test

jkakavas commented:
ping @tvernum in case you want to take another look since I also cleaned up the SamlAuthenticationState and added a field to the response as we talked about

jkakavas commented:
backport to 7.7 depends on https://github.com/elastic/elasticsearch/pull/54553 being merged
